### DISCUSSION TITLE
Vote on verification build warning reporting

### DISCUSSION THREAD

**Discussion by akurtakov** [eclipse-platform/eclipse.platform.ui#1567] [Polls]
https://github.com/eclipse-platform/eclipse.platform.ui/pull/1557 lead to discussion that shows lack of agreement in where/how/why/etc. the community wants to see warnings in verification builds. Please read the comments in the PR for extra information

**Comment by mickaelistria** [comment]
FWIW, I find warnings are only good in the IDE while developing because they help making code better as you work on it. So I voted for disabling reporting all together at build time as I don't find them helpful at that time.

**Reply by laeubi** [reply]
> FWIW, I find warnings are only good in the IDE while developing because they help making code better as you work on it.

The problem is that

1. JDT does not detect all issues in javadoc in the IDE: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1858
2. If its only a warning often people probably don't care, probably never fix and they accumulate more and more

Because of that it was asked to enable / make them visible in the CI so a reviewer can get aware of it without having to start the IDE and look through thousands of warnings to spot one new item.

**Reply by mickaelistria** [reply]
While https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1858 is related in the story here, in the end, it's a separate issue/fix to add in JDT (or any other static analyzer we'd like to hook in the IDE or build).

> If its only a warning often people probably don't care, probably never fix and they accumulate more and more

Yes, but the very reason why warnings exists instead of errors are that often they're good just being "only a warning" and allowing people to not care too much about them (ie not block). If we want people to fully care and turn some warnings into blockers, then those warnings needs to be reported as errors in the IDE and build.

**Reply by laeubi** [reply]
And that's where the circle closes, because we getting more and more warnings and "no one" will fix all of them we can't turn them into errors, because people thing they are "just warnings" ;-)

As a side note, these tools does not only report **warnings** they also report errors and the main benefit is that a user do not need to scan/parse the log itself, so if there is a API **error** or compile **error** or whatever it is easily accessible.

**Comment by vogella** [comment]
Loosely related: I think we should enable that the build fails if the number of warnings increases. IIRC @laeubi had a solution for this, but did dare to enable it, as this may also break the build in other cases, e.g. if the compiler suddenly adds more warnings.

**Reply by laeubi** [reply]
Just wanted to not that even though the blame is always put on me, it is not that **I (personally)** have a solution (or demand for them), but these are _software industry standard tools_ (namely **Jenkins** + **Plugins** (according to their stats [warnings-ng](https://github.com/jenkinsci/warnings-ng-plugin) has 23k installs) and yes they offer such (standard) options.

Also in the past it was often complained that it is "too hard to find the issue of the build failure" so we installed this **with approval by the project leads** of the projects so we can at least show them in a more structured way than a lengthy build log.

**Reply by vogella** [reply]
My intention was to **give you credit for your great work** as you learn about and apply the _software industry standard tools_ to the Eclipse project which unfortunately is not using these standard tools. My blame goes usually to people which do not want to apply these tools or apply the the *why change things, we always did it like this* approach.

**Reply by laeubi** [reply]
@vogella we are all standing on the shoulders of giants, so my work is just small compared to all the work done before in these tools and plugins, so thanks for the credits but I actually only applied the configuration and I'm happy that is already works to such extend but far from that I crafted a solution I just applied what was already there!

So from my side its now up to us to give feedback to the authors of these great tools if something needs improvement instead of complain its not 100% fit for our purpose... anyone can improve its even open source.

**Reply by vogella** [reply]
I agree that it is great that these tools are developed as OS. 

In the context of Eclipse it is still worth giving gratitude to people which still fighting the fight of applying industrial standards to Eclipse. In the past the "not invented here" and "why change something" mindset was very present and the burden of this mindset is in my opinion still present in the project.  Fortunately the project team is much more open to changes these days. But I will stop now posting not closely unrelated topics into @akurtakov  voting.

**Comment by jukzi** [comment]
Now all jdt master builds are rated unstable, because ECJ reports "Quality gate: Unstable". If anybody volunteers to fix that - welcome. otherwise i will even disable ECJ again
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/51790620/12ae2b69-05d2-4b35-930c-1615d130fdf1)
pressing "Reset" did not help. The ECJ plugin seems to find but ignore build#30 ff.
```
14:24:33  [Eclipse ECJ] Obtaining reference build from reference recorder
14:24:33  [Eclipse ECJ] -> Found 'eclipse.jdt.core-Github » master #30'
14:24:33  [Eclipse ECJ] Using reference build 'eclipse.jdt.core-Github/master #27' to compute new, fixed, and outstanding issues
```

**Reply by akurtakov** [reply]
@jukzi Which build is that? Have you tried to use the Reset button? 
It's really important that we learn to respect each other's opinion more. So far "disable none" seems to be preferred vote and everyone is supposed to respect majority opinion (despite not being yours or mine vote).

**Reply by jukzi** [reply]
> pressing "Reset" did not help. 

https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/master/

**Reply by akurtakov** [reply]
So 
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/574788/febba334-c2cf-4c6b-8915-7f9bf263b522)
means that next build will use that one as a base for comparison aka warnings that exist in it shouldn't flag the build and indeed https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/master/32/ is green again.

**Reply by jukzi** [reply]
"the next build" was not true. we had dozen of PR builds after "Reset" which did not move its reference.
So the workaround was to "Reset"+ manually retrigger master? Someone volunteer for all other repos too, please. Or can that be somehow automated? Like "ignoreQualityGate: true" (for master only?)

**Comment by iloveeclipse** [comment]
Here is another [unrelated error reporting](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928/checks?check_run_id=21026374652) that now fails JDT builds for no obvious reason (PR touched **different bundle**): https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928.

So please can we stop this experiments? I don't see an end on **unrelated** errors reported. They may be valid for some, but they are highly distracting. Every PR fails and very time I check the failure it is something unrelated to the PR.

**Reply by akurtakov** [reply]
For the record - this vote I started here is for Platform (even UI) as the PR that started the discussion is against this repo. 
I would really like to see projects being more standalone as what is best for Platform is not best for JDT or vice-versa, furthermore there is a good amount amount of people that are committers on only one of the projects. 
So if Project A comes to a decision that the project wants every possible warning shown and Project B decides that they would like to see no warning at all - it's project's committers decision.

**Reply by akurtakov** [reply]
If there is disagreement about projects deciding on their own that we should be looking into merging everything into single project (for the record I'm strongly against that!) so there is no case where the need of one subproject puts more/less policies on the other one.

**Reply by iloveeclipse** [reply]
My comment was not JDT specific, I'm tired to see red state on PR's with unrelated changes, and that is for **all** repositories we have in SDK. JDT PR was simply the latest ome I saw... Every time there is a different explanation why this particular PR is failing, and that has to find an end. 

We should disable all the unstable checks and only enable them if they would provide reasonable and reproducible results.

**Reply by akurtakov** [reply]
What I try to point is - you're free to do so for JDT (if a committer disagrees he/she should should start a vote in JDT). So far the Platform committers seems to prefer the other way around and see more warnings and look into whether they make sense to them, fix them overtime, ignore and etc. 
Such a diversity between projects might actually be best for everyone as it should help us see which one gives better results over some longer(months?) period.

**Reply by akurtakov** [reply]
Speaking of failing builds - last good aggregator verification master build was on October 31st https://ci.eclipse.org/platform/job/eclipse.platform.releng.aggregator/job/master/ without any of these extra warnings and etc. being enabled and no one noticed! So clearly we need things to change but the disagreement is what - make more things visible everywhere so more people see it and hopefully fix things or hide things and pretend everything is fine.
 In the end it's the very same small group of people fixing things and after being the one suffering from javadoc build failures for years, I consider enabling them for everyone so people went and fix them (HUGE THANKS FOR THAT!!!) a total win.

**Comment by laeubi** [comment]
> Here is another [unrelated error reporting](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928/checks?check_run_id=21026374652) that now fails JDT builds **for no obvious reason** (PR touched different bundle)

I really fail to understand how one could make it more obvious whats the problem is

![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/1331477/492cf676-3fbb-40ba-bcb4-327ab484ac8f)

Fixing these 4 API Warnings as part of the PR (or even a separate one if one cares) would probably be much more easier than write the comment and getting mad again. Also @akurtakov has explained already to how reset the quality gate for master if one really things it is to much effort to improve the warnings just a little bit, that might be a bit annoying as JDT has so many different repositories but this is [the decision of the team to not act on it](https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/615) and instead higher the releng burden, please note **I can't do it** because I'm not a committer JDT!

Also as @akurtakov has mentioned JDT is a separate project and can take distinct decisions, but it should then also take the responsibility for the releng work, e.g. this specific issue you mentioned is caused by the fact that JDT still seem not perform the necessary actions (read: **ignores the "unreleated" warning!**) that a baseline replace has happened on a bundle with different content resulting in 
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1757 (would have been discovered directly on PR time)
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1750 (would have be discovers if not using many different repositories)
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1704 (would have been discovered directly on PR time)
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1686 (would have be discovers if not using many different repositories)
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1659 (would have been discovered directly on PR time)
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1617 (would have been discovered directly on PR time)

**Reply by iloveeclipse** [reply]
> this specific issue you mentioned is caused by the fact that JDT still seem not perform the necessary actions (read: ignores the "unreleated" warning!) that a baseline replace has happened

The issue I was fixing is **NOT** related to that at all! It is caused by constants value change in a different bundle. And the warnings I've got on PR were **unrelated** to that change.

> but it should then also take the responsibility for the releng work

- **Every single problem** you've mentioned was handled by "JDT team".
- All of them are outside of JDT repository in different platform repositories, so surely JDT can't be blamed to fix compiler bugs **and** **also** do a releng work after that?
- There is a "help wanted" ticket to improve that https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1917 so whoever has enough maven/tycho knowledge is welcome to contribute
- Not only that, most of other platform build issues and warnings were fixed by "JDT team".

I dislike the attitude to separate "good teams" from "bad teams" that I see here and saying "the good ones" always do the right thing and "the bad ones" only complain. It is not correct and does not true.

All what Jörg and me ask here is to disable **unrelated** warnings on PR's and that alone is obviously enough to put us to the "bad team". So well, I'm happy to be in that team...

**Reply by laeubi** [reply]
None of the mentions ibuild failures where needing the "help wanted" ticket, they would have been discovered already right now (for the jdt part of course not where a larger audience is also affected).

Also one can't "disable unrelated warnings" as a tool can't know what a developer in a certain situation assumes as "unrelated", deltas are compare against reference build, so if reference build is not correct (as it was not run yet) it will assume they are new, how else would it work? For example Eclipse is showing me all the day thousands of "unrelated" warnings in Platform code all not related to the file I'm currently changing a single line... that's of course perfectly fine, but a CI tool on the other hand now should magically "know" what relates to a change?!

**Reply by iloveeclipse** [reply]
> None of the mentions ibuild failures where needing the "help wanted" ticket, they would have been discovered already right now

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1917 **is required** to detect that, and I don't know other solution.

Please share how one can fix a difference between class files compiled with compiler version A that differ from classes compiled with compiler version B for classes that are built in different repositories?

**Reply by merks** [reply]
Indeed, perhaps you are both expecting the impossible of the other and are not being very nice about it to boot.

I've really, really, really had enough of finger pointing, blaming, and griping.  It's beyond tiresome and it is beneath the decorum of what one would expect from such highly-skilled, technical experts who are all working so incredibly hard to make Eclipse excellent. You should all be thanked by the masses every single day.  The community respects you folks and appreciate what you do even if so few ever bother to say so.

In any case, the build is not an AI engine that understand related versus unrelated problems.  It can check if the problem existed historically and filter those as noise.  That is what the releng folks are trying to provide.   Also, when we build with a new compiler every day, we will notice when it produces different byte code regularly.   That will always be causing us problems.  It does little good for me to say, please make that problem go away as if a magic wand could be used.   Better to ask, what can we do to mitigate that problem?  Is there some additional testing we can do? 

Please, please dear goodness, could we focus on what can be improved and how best to improve it without some folks being bad folks and some folks being good folks?

**Reply by laeubi** [reply]
> Please share how one can fix a difference between class files compiled with compiler version A that differ from classes compiled with compiler version B for classes that are built in different repositories?

Please read below **just as a hint to give advice how this can be detected as of today**, I'm not blaming anyone, and just want to highlight _how_ it works and _why_ it works and what was meant with 

>  this specific issue you mentioned is caused by the fact that JDT still seem not perform the necessary actions (read: ignores the "unreleated" warning!)

and why it is actually bad to qualify anything as "unrelated" just because it get into the way to press submit button, even worse [this ticket was blamed there](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928#issuecomment-1917452789) to be "Validation error as usually unrelated" even though this issue where never came up to the integration build at all if the warnings where taken as what they are **a hint to the developer that something might go wrong**.

## Issue that was detected in the I-Build later

Here
- https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/1757

looking into the [buildtimeComparatorUnanticipated.log.txt](https://download.eclipse.org/eclipse/downloads/drops4/I20240129-1800/buildlogs/comparatorlogs/buildtimeComparatorUnanticipated.log.txt) you will see the this message

```
1.  eclipse.jdt.core/org.eclipse.jdt.core.tests.performance/pom.xml
   no-classifier: different
      org/eclipse/jdt/core/tests/performance/FullSourceWorkspaceBuildTests.class: different
```

##  causing change and verifications

You referenced this change
- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1517

When I look at the build verification it shows
![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/1331477/3c4207fb-4836-4d29-896b-dceedf174daf)

clicking on the details of this new warning it shows

![grafik](https://github.com/eclipse-platform/eclipse.platform.ui/assets/1331477/152273c9-49c5-4d7b-9278-8af82e665fea)

> MavenProject: org.eclipse.jdt:org.eclipse.jdt.core.tests.performance:3.12.300-SNAPSHOT @ /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1517/org.eclipse.jdt.core.tests.performance/pom.xml: baseline and build artifacts have same version but different contents

## Conclusion

As you can see it is the same project mentioned here (the full context can't be viewed anymore because the build is already deleted), and one could have made the touch of the bundle part of the PR as you have done after the fact here

- https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928

then the warning would have been fixed in the verification build, the I build would have run without an issue, one could have looked at the result if it is maybe an undesired side effect before merging.

**Reply by iloveeclipse** [reply]
> Here is another [unrelated error reporting](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928/checks?check_run_id=21026374652) that now fails JDT builds for no obvious reason (PR touched different bundle): https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1928.

Once again, the warnings **I've** mentioned above were **unrelated** to the PR's in question. 
I'm not responsible if anyone else ignores some warnings on other PR's and I can also understand why it happens.

> None of the mentions ibuild failures where needing the "help wanted" ticket

This is still wrong, and there is still no solution proposed for that except https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1917.
I'm not sure why you are denying to see the obvious. The only example you've shown was "per occasion" in same repository.
For compiler related SDK build issues I do not see any other solution as to provide a fix for https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1917.

And yes, there might be sometimes "lucky" PR validation errors reported that could be relevant, but they are being simply ignored by almost everyone because there are **always** some validation errors, and that is the problem we are trying to explain here.

**Reply by jukzi** [reply]
> /home/jenkins/agent/workspace/eclipse.jdt.core-Github_PR-1517/org.eclipse.jdt.core.tests.performance/pom.xml: baseline and build artifacts have same version but different contents

Its cool if we could prevent such errors from to be merged. And this explanation makes clear what should be done.: It would be even better if this error message would 
1. explicitly rate it as error - given that it will fail later builds. 
2. give an explanation why this is an issue and what has to be done, as a normal java developer does not have the knowledge to understand and solve the error.

**Comment by jukzi** [comment]
Let's not put mental barrier between different teams. At the end we have the common goal to contribute to the eclipse IDE.
The skillset between the people is very mixed. Not everybody has the skill or time to care about issues in topics like build/javadoc/API/compiler/UI/Multithreading/memory/performance/UX/... .It's a high value to have different skills. The tools should not force everybody to care about the same issue. Please respect that not everybody has the same focus. If anyone is able to fix any issue it is highly welcome in all projects. 
For example i don't want to put focus on the releng things and therefore i don't want to get alerted about releng issues, even if i could solve them. I am very happy that people like @akurtakov and @laeubi do work on it.  Just please keep the tools in a way that no one else needs to care about, because others can't or just won't.  I don't want everybody to need to know how to reset quality gate. Either the people that want that gate enabled please provide the necesessary time to manually monitor and reset it  or automate it.

> Fixing these 4 API Warnings as part of the PR (or even a separate one if one cares) would probably be much more easier

Please provide a PR - i don't know how to fix that off the cut.

> decision of the team to not act on it

Please respect that. Making releng easier is one good purpose, but when even more people suffer from merging repos, for example by higher build times it's also a legit perspective where we need to find a compromise between goals.

**Reply by laeubi** [reply]
> Please provide a PR - i don't know how to fix that off the cut.

Know / accepted problems can be filtered and you already have done this in the past so I can't really believe that but for reference just in case: https://wiki.eclipse.org/PDE/API_Tools/User_Guide#Filtering_Known_Problems

Filters that are no longer being used can simply be delete, in both cases Eclipse offers you a Quickfix you can use.

>  Just please keep the tools in a way that no one else needs to care about, because others can't or just won't. 

or just don't want ... it is the responsibility of a committer to care there is no way to shift responsibilities to a random person.

**Reply by merks** [reply]
@jukzi 

I'm not sure I can agree with this line or reasoning, and certainly not when taken to the extreme.  If I don't care about warnings in my Java code and then contribute/modify code that introduces a bunch of new warnings, should someone else who cares be responsible for fixing them?  I would hope and expect not.  If I don't care that the javadoc *that I do bother to write) is full of errors, should I just submit it anyway and expect someone who cares to fix it?  Surely we're all responsible for caring to a significant extent about the overall quality of what we produce collectively.  I expect that subsets of people not caring about subsets of issues is what leads to an overall union of problems so numerous that not caring appears to be the only way to cope.  After all, one cannot be all too concerned about adding or removing  warning from this sea of warnings:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/b38dabf2-b97e-4fd8-8a90-4fdf13c0d1c8)

In my projects, warnings are eliminated and new ones are noticed immediately:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/a6ff723a-b98b-46b0-9525-a9165e7d5c8f)

One thing that I can say with certainty is that I get floods of email every day and all too often lately I **do not like** the adversarial tone that's taken.  It's negative and depressing.  It makes me want to ignore this like ignoring warnings or to go somewhere where there are nice people to work with.  When the word "please" takes on the tone of a demand, you know that communication has degraded down the wrong path.

Do you think it's possible that we focus primarily on improving what we have?  For example, personally I think it would be great to have a central location where someone who is uninformed can inform themselves on how to interpret the build results and how to find details about problems.  Looking at recent example:

https://github.com/eclipse-pde/eclipse.pde/pull/1094

it appears to be working well when it's working well:

![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/208716/e857408d-cefa-49ed-9d66-f9e0d477b63a)

Oh wow, and someone saying thank you.  How nice is that?

**Reply by jukzi** [reply]
We agree about the responsibility is on those who introduce warnings. But its complicated: who is responsible for 
* code that date back to committers gone
* new warnings introduced by 3rd party changes (deprecation, new features like generics ...)
* new warnings by activating new warning tools, changing existing tools, reconfiguring existing tools?

Does it help to know who is responsible? - I fixed thousands (!) of warnings in the last months that i am not responsible for including those shown by no tool at all. The new tools did not help with that, instead i had to take extra time to fix  or ignore those. 

Please don't take this discussion personally, its only about the thing not the people. I only try to find way to get happy with the toolchain.

**Reply by akurtakov** [reply]
You fixing all these warnings allowed us to move ahead and further stabilize a lot - huge thanks! 
Responsibility for the state of the codebase is shared between all committers (some do more than others but that's FOSS by nature) and every single one of us is responsible. 

The goal with all these extra checks is not to make committer's life harder rather the opposite - catch issues earlier and make them seen earlier and eventually prevent from new ones entering the codebase. Unfortunately, the current state is such that if we just hide warnings more will enter and getting more automation/checks will become even harder to achieve in the future.

The only exit from this vicious cycle seems to be to enable more till we fix, raise from warning to error, require extra approvals for importing new ones and so on and on. It is a painful process for sure but the other option seems to be to accept things degrading . 
Github/Jenkins provide us with great way for more checks and automations and we are barely scratching the surface of possibilities now. E.g. see what we are trying in Tycho now (https://github.com/eclipse-tycho/.eclipsefdn/commit/42c0aa8269368bb12ae3afc73c99ff59e3780e4b#diff-88861c0250c496da5543bd28a02b2e3cf88129271a4731b2c6dd5e92a76712f4R82) to mandate that verification build and license check are prereqs, hopefully we get to a point when Platform codebase gateway is hardened in such way too so we see less sudden breakage. For me one thing is sure - if we don't try we will never get there and yes, there will be (a lot of!) growing pain.

**Comment by jukzi** [comment]
From 9 votes 5 votes to disable maven, 4 voted to keep maven.  
Only 3 voted to disable javadoc.
So lets disable maven, keep javadoc?

**Reply by akurtakov** [reply]
Lesson for myself and others for next votes: All votes should be binary so there is no room for interpretation!
The only way multi option votes can work are if we have https://en.wikipedia.org/wiki/Score_voting which I don't think we can achieve via github thus I'll create a new binary vote for disabling Maven so there could be no question about integrity of the voting process and interprecation of results.

**Reply by akurtakov** [reply]
I've started a new one for the sake of formalizing and preventing any possible miscommunication.

**Reply by akurtakov** [reply]
https://github.com/eclipse-platform/eclipse.platform.ui/discussions/1644

**Comment by akurtakov** [comment]
Close due to non-clear result. To be clarified in https://github.com/eclipse-platform/eclipse.platform.ui/discussions/1644.
